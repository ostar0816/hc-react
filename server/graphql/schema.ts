import { GraphQLDateTime, GraphQLDate } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';
import StudyTableColumns from '../constants/studyTableColumns';
import { resolveViewerType, GraphQLObjectId } from './helpers';
import { gql } from 'apollo-server';

import { viewerResolvers } from './resolvers/viewer';
import { userResolvers } from './resolvers/user';
import { studyResolvers } from './resolvers/study';
import { worklistResolvers } from './resolvers/worklist';
import { studyDescriptionResolvers } from './resolvers/studyDescription';
import { facilityResolvers } from './resolvers/facility';
import { studyNoteMessageResolvers } from './resolvers/studyNoteMessage';

import { mutationResolvers } from './resolvers/mutation';

import { GQLResolver } from './types';

export const typeDefs = gql`
  scalar DateTime
  scalar Date
  scalar JSON
  scalar ObjectId

  enum PermissionEnum {
    LIST_ALL_STUDIES
    LIST_ALLOWED_STUDIES
    LIST_STUDIES_ASSIGNED_TO_ME
    ASSIGN_STUDY
    CONFIRM_STUDY
    OPEN_REPORT_WINDOW
    SIGN_OFF_STUDY
    RESOLVE_SUPPORT_REQUEST
  }

  enum ContrastEnum {
    NA
    W
    WO
    W_WO
  }

  enum ContrastPresentEnum {
    NOT_SELECTED
    YES
    NO
  }

  enum ContrastRequirementsEnum {
    NEVER_SEEN
    OPTIONAL
    REQUIRED
  }
  
  enum ModalityEnum {
    CT
    MR
    CR
    DX
    XR
    US
    PET
    MG
    NM
  }

  enum ReadTypeEnum {
    PRELIM
    PRELIM_OR_FINAL
    FINAL
  }

  enum StudyStatusEnum {
    NEW
    CONFIRMED
    ASSIGNED
    SIGNED_OFF
  }

  enum StudyPriorityEnum {
    STAT
    URGENT
    CRITICAL
    ROUTINE
    STROKE
  }

  enum StudyLateralityEnum {
    NOT_SELECTED
    LEFT
    RIGHT
  }

  enum AssignPriorityEnum {
    ZERO
    PLUS_ONE
    PLUS_TWO
    MINUS_ONE
    MINUS_TWO
    NEVER
  }

  enum NoteTypeEnum {
    SUPPORT_STUDY_REQUEST
    SUPPORT_URGENCY_REQUEST
  }

  enum StudySupportRequestStatusEnum {
    NONE
    PENDING
    RESOLVED
  }

  type AssignOption {
    radiologist: User!
    assignPriority: AssignPriorityEnum!
  }

  type Study {
    _id: ObjectId!
    canBeAssignedTo: [AssignOption]
    assignedTo: User
    contrastPresent: ContrastPresentEnum
    contrastType: String
    studyDescription: StudyDescription
    facility: Facility
    referringPhysicians: [ReferringPhysician]
    priorStudies: [Study]!
    notes: [StudyNote]
    report: Report
    addendums: [Report!]!
    ${Object.keys(StudyTableColumns).map(
      columnKey => `${columnKey}: ${StudyTableColumns[columnKey].type}`,
    )}
  }
  
  type StudySearchResult {
    totalCount: Int!
    studies: [Study]
  }

  enum FilterEnum {
    CONTAINS_STRING
    EQUAL_NUMBER
    DATE_RANGE
    DATE_PRESET
  }

  input SearchFilterInput {
    columnName: String!
    type: FilterEnum!
    stringValue: String
    numberValue: Float
    startDate: Date
    endDate: Date
    datePreset: String
  }
  
  input SearchSortInput {
    columnName: String!
    order: Int
  }

  type StudyNoteMessage {
    createdBy: User!
    createdTime: DateTime
    text: String
  }

  type StudyNote {
    _id: ObjectId!
    type: NoteTypeEnum
    title: String!
    messages: [StudyNoteMessage]!
    resolved: Boolean
  }

  type Filter {
    type: FilterEnum!
    stringValue: String
    numberValue: Float
    startDate: Date
    endDate: Date
    datePreset: String
  }

  type Column {
    key: String!
    width: Int!
    filter: Filter
  }

  type SortType {
    order: Int!
    columnName: String!
  }

  type StudyDescription {
    _id: ObjectId!
    name: String!
    modality: ModalityEnum
    contrast: ContrastEnum
    reportTemplate: ReportTemplate
  }
  
  type ReportTemplate {
    _id: ObjectId!
    name: String!
    description: String!
    template: JSON!
  }

  type Report {
    content: JSON!
    signedOffDate: DateTime
  }


  type Worklist {
    _id: ObjectId!
    name: String!
    displayOnTop: Boolean!
    description: String
    sorting: [SortType]
    columns: [Column]
    totalCount: Int
  }
  
  type Workspace {
    _id: ObjectId!
    layout: JSON
  }

  type AssignPriority {
    studyDescriptionNameRegex: String!
    studyDescriptionModalityRegex: String!
    facilityNameRegex: String!
    studyPriorityRegex: String!
    assignPriority: AssignPriorityEnum!
  }
  
  type User {
    _id: ObjectId!
    firstName: String
    lastName: String
    email: String!
    workspaces: [Workspace]
    worklists: [Worklist]
    userRole: UserRole
    assignedFacilities: [Facility]!
    allowedFacilities: [Facility]!
    assignPriorities: [AssignPriority]
    pendingStudies: [Study]
    online: Boolean!
    radiologyCapacity: Int
  }
  
  type UserRole {
    _id: ObjectId!
    name: String!
    description: String!
    permissions: [PermissionEnum]
  }
  
  type ReferringPhysician {
    _id: ObjectId!
    physicianUid: String!
    name: String!
    email: String!
    phone: String!
    fax: String
    dicomValue: String
  } 
  
  type FacilityDicomDetails {
    AETitle: String
  }
  
  type TATLimit {
    timeSeconds: Int!
    studyPriority: StudyPriorityEnum!
  }

  input TATLimitInput {
    timeSeconds: Int!
    studyPriority: StudyPriorityEnum!
  }

  type FacilityTechConf {
    contrastRequirements: [ContrastRequirement!]!
    readType: ReadTypeEnum
    tatLimits: [TATLimit!]!
  }
  
  type Facility {
    _id: ObjectId!
    institutionName: String!
    address: String
    city: String
    state: String
    zip: String
    websiteUrl: String
    faxNumber: String
    phoneNumber: String
    email: String
    dicomDetails: FacilityDicomDetails
    techConf: FacilityTechConf
    referringPhysicians: [ReferringPhysician]
  }
  
  
  type Viewer {
    _id: ObjectId!
    user: User
    studyDescriptions: [StudyDescription]
    facilities: [Facility]
    reportTemplates: [ReportTemplate]
    userRoles: [UserRole]
    users: [User]
    searchFacilityById(facilityId: ObjectId): Facility
    searchStudies(skip: Int!, limit: Int!, sorting: [SearchSortInput], filters: [SearchFilterInput]): StudySearchResult
    searchStudyById(studyId: ObjectId): Study
    searchUserById(userId: ObjectId): User

  }
  
  type Query {
    viewer: Viewer
  }
  
  input ColumnInput {
    key: String!
    width: Int!
    filter: FilterInput
  }
  
  input FilterInput {
    type: FilterEnum!
    stringValue: String
    numberValue: Float
    startDate: Date
    endDate: Date
    datePreset: String
  }
  
  input SortInput {
    columnName: String!
    order: Int!
  }

  input ContrastRequirementsInput {
    modality: ModalityEnum!
    requirement: ContrastRequirementsEnum!
  }

  type ContrastRequirement {
    modality: ModalityEnum!
    requirement: ContrastRequirementsEnum!
  }
  
  input UserRolePermissionsInput {
    userRoleId: ObjectId!
    permissions: [PermissionEnum]!
  }

  input AssignPriorityInput {
    studyDescriptionNameRegex: String!
    studyDescriptionModalityRegex: String!
    facilityNameRegex: String!
    studyPriorityRegex: String!
    assignPriority: AssignPriorityEnum!
  }


  type Mutation {
    login (
      email: String!
      password: String!  
    ) : Query
    
    logout : Query
    
    addWorklist (
      name: String!
      description: String!
      sorting: [SortInput]!
      columns: [ColumnInput]!
    ): Query
    
    updateWorklist (
      worklistId: ObjectId!
      sorting: [SortInput]!
      columns: [ColumnInput]!
    ): Query

    updateWorklistDisplayOnTop (
      worklistId: ObjectId!
      displayOnTop: Boolean!
    ): Query

    deleteWorklist (
      worklistId: ObjectId!
    ) : Query

    updateWorkspace (
      workspaceId: ObjectId!
      layout: JSON!
    ) : Query
    
    addStudyDescription (
      name: String!
      modality: ModalityEnum!
      reportTemplateId: ObjectId
      contrast: ContrastEnum!
    ) : Query
    
    deleteStudyDescription (
      studyDescriptionId: ObjectId!
    ) : Query
    
    updateStudyDescription (
      studyDescriptionId: ObjectId!
      name: String!
      modality: ModalityEnum!
      contrast: ContrastEnum!
      reportTemplateId: ObjectId
    ) : StudyDescription
    
    addReportTemplate (
      name: String!
      description: String!
      template: JSON!
    ) : Query
    
    updateReportTemplate (
      reportTemplateId: ObjectId!
      name: String!
      description: String!
      template: JSON!
    ) : ReportTemplate
    
    addUserRole (
      name: String!
      description: String!
    ) : Query
    
    updateUserRole (
      userRoleId: ObjectId!
      name: String!
      description: String!
    ) : UserRole

    addUser (
      firstName: String!
      lastName: String!
      email: String!
      password: String!
      userRoleId: ObjectId!
    ) : Query
    
    updateUser(
      userId: ObjectId!
      email: String!
      firstName: String!
      lastName: String!
      password: String
      userRoleId: ObjectId!
    ) : User
  
    updateUserRolePermissions(
      userRolePermissions: [UserRolePermissionsInput]
    ) : Query

    deleteReportTemplate (
      reportTemplateId: ObjectId!
    ) : Query

    addFacility (
      institutionName: String!
      address: String!
      city: String!
      state: String!
      zip: String!
      websiteUrl: String!
      faxNumber: String!
      phoneNumber: String!
      email: String!
    ) : Query
    
    updateFacility (
      facilityId: ObjectId!
      institutionName: String!
      address: String!
      city: String!
      state: String!
      zip: String!
      websiteUrl: String!
      faxNumber: String!
      phoneNumber: String!
      email: String!
    ) : Facility
    
    deleteFacility (
      facilityId: ObjectId!
    ) : Query
    
    addReferringPhysician (
      facilityId: ObjectId!
      physicianUid: String!
      name: String!
      email: String!
      phone: String!
      fax: String!
      dicomValue: String!
    ) : Facility
        
    updateReferringPhysician (
      referringPhysicianId: ObjectId!
      physicianUid: String!
      name: String!
      email: String!
      phone: String!
      fax: String!
      dicomValue: String!
    ) : ReferringPhysician
        
    deleteReferringPhysician (
      referringPhysicianId: ObjectId!
    ) : Facility
        
    updateFacilityTechConf (
      facilityId: ObjectId!
      contrastRequirements: [ContrastRequirementsInput]!
      readType: ReadTypeEnum!
      tatLimits: [TATLimitInput!]!
    ) : Facility
    
    updateFacilityDicomDetails (
      facilityId: ObjectId!
      AETitle: String!
    ) : Facility
    
    confirmStudy(
      studyId: ObjectId!
      patientName: String!
      patientId: String!
      patientDOB: Date!,
      patientSex: String!
      studyDescriptionId: ObjectId!
      contrastPresent: ContrastPresentEnum!
      contrastType: String!
      studyPriority: StudyPriorityEnum!
      laterality: StudyLateralityEnum!
      readType: ReadTypeEnum!
      history: String!
      referringPhysicianIds: [ObjectId!]!
    ) : Study

    updateAssignPriorities(
      userId: ObjectId!
      assignPriorities: [AssignPriorityInput]!
    ) : User

    assignStudiesToRadiologist(
      radiologistId: ObjectId!
      studyIds: [ObjectId!]!
    ) : User

    updateAllowedFacilitiesForUser(userId: ObjectId!, facilityIds: [ObjectId!]!) : User
    signOffReport(studyId: ObjectId!, content: JSON!) : Study
    signOffAddendum(studyId: ObjectId!, content: JSON!) : Study

    addStudyNote(studyId: ObjectId!, title: String!, text: String!, type: NoteTypeEnum) : Study
    addStudyNoteMessage(studyId: ObjectId!, noteId: ObjectId!, text: String!, resolved: Boolean) : Study

    updateStudyFacility(studyIds: [ObjectId!]!, facilityId: ObjectId!) : [Study!]!
  }

`;

const queryResolvers = {
  viewer: resolveViewerType,
};

export const resolvers: GQLResolver = {
  Query: queryResolvers,
  Viewer: viewerResolvers,
  User: userResolvers,
  Study: studyResolvers,
  Worklist: worklistResolvers,
  Facility: facilityResolvers,
  StudyDescription: studyDescriptionResolvers,
  StudyNoteMessage: studyNoteMessageResolvers,
  Mutation: mutationResolvers,
  DateTime: GraphQLDateTime,
  Date: GraphQLDate,
  JSON: GraphQLJSON,
  ObjectId: GraphQLObjectId,
};
