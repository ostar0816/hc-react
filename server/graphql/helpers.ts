import { GQLViewer } from './types';
import { ObjectId } from 'mongodb';
import { GraphQLScalarType } from 'graphql';
import { Kind } from 'graphql/language';

export const GraphQLObjectId = new GraphQLScalarType({
  name: 'ObjectId',
  description: 'Id representation, based on Mongo Object Ids',
  parseValue(value) {
    return new ObjectId(value);
  },
  serialize(value) {
    return value.toString();
  },
  parseLiteral(ast) {
    if (ast.kind === Kind.STRING) {
      return new ObjectId(ast.value);
    }
    return null;
  },
});

export function resolveViewerType(): GQLViewer {
  return {
    _id: '1',
  };
}
