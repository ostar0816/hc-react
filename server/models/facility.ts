'use strict';
import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import {
  GQLReadTypeEnum,
  GQLContrastRequirement,
  GQLModalityEnum,
  GQLContrastRequirementsEnum,
  GQLPriorityTimeRequirement,
  GQLTATLimit,
} from '../graphql/types';
import { GQLContext } from '../utils/getGraphqlContext';

export type MFacilityBase = {
  _id: ObjectId;
  institutionName: string;
  address: string;
  city: string;
  state: string;
  zip: string;
  websiteUrl: string;
  faxNumber: string;
  phoneNumber: string;
  email: string;
};

export type MFacilityTechConf = {
  readType: GQLReadTypeEnum;
  contrastRequirements: Array<GQLContrastRequirement>;
  tatLimits: Array<GQLTATLimit>;
};

export type MFacilityDicomDetails = {
  AETitle: string;
};

export type MFacility = MFacilityBase & {
  techConf: MFacilityTechConf;
  dicomDetails: MFacilityDicomDetails;
  assignedRadiologistIds: Array<ObjectId>;
};

export function getFacilityDefaults() {
  return {
    techConf: {
      readType: GQLReadTypeEnum.PRELIM,
      contrastRequirements: [
        {
          modality: GQLModalityEnum.CT,
          requirement: GQLContrastRequirementsEnum.REQUIRED,
        },
        {
          modality: GQLModalityEnum.MR,
          requirement: GQLContrastRequirementsEnum.REQUIRED,
        },
      ],
      tatLimits: [],
    },
    dicomDetails: {
      AETitle: '',
    },
    assignedRadiologistIds: [],
  };
}

export function cFacility() {
  return getDb().collection<MFacility>('facility');
}

export async function listFacilities(ctx: GQLContext) {
  checkPerms(ctx);

  return cFacility()
    .find({})
    .sort([['institutionName', 1]])
    .toArray();
}

export async function getFacilitiesByAssignedPhysicianId(ctx, userId: ObjectId) {
  checkPerms(ctx);

  return cFacility()
    .find({ assignedRadiologistIds: userId })
    .sort([['institutionName', 1]])
    .collation({ locale: 'en', strength: 2 })
    .toArray();
}

export async function getFacilityById(ctx, facilityId: ObjectId): Promise<MFacility> {
  checkPerms(ctx);

  return ctx.loaders.Facility.load(facilityId);

  //return Facility.findOne({_id: facilityId}).exec();
}

export async function loadFacilityByIds(ctx, facilityIds: ObjectId[]): Promise<MFacility[]> {
  checkPerms(ctx);

  return ctx.loaders.Facility.loadMany(facilityIds);

  //return Facility.findOne({_id: facilityId}).exec();
}

export async function addFacility(ctx: GQLContext, facility: Omit<MFacilityBase, '_id'>) {
  checkPerms(ctx);

  return cFacility().insertOne({ ...getFacilityDefaults(), ...facility });
}

export async function updateFacility(
  ctx,
  facilityId: ObjectId,
  facility: Omit<MFacilityBase, '_id'>,
) {
  checkPerms(ctx);

  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    {
      $set: facility,
    },
    { returnOriginal: false },
  )).value;
}

export async function updateFacilityTechConf(
  ctx: GQLContext,
  facilityId: ObjectId,
  facilityTechConf: MFacilityTechConf,
) {
  checkPerms(ctx);

  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    { $set: { techConf: facilityTechConf } },
    { returnOriginal: false },
  )).value;
}

export async function updateFacilityDicomDetails(
  ctx: GQLContext,
  facilityId: ObjectId,
  facilityDicomDetails: MFacilityDicomDetails,
) {
  checkPerms(ctx);
  console.log(facilityId, facilityDicomDetails);
  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    {
      $set: {
        dicomDetails: facilityDicomDetails,
      },
    },
    { returnOriginal: false },
  )).value;
}

/*
const Facility = new Schema(
  {
    institutionName: { type: String, unique: true },
    address: { type: String, default: '' },
    city: { type: String, default: '' },
    state: { type: String, default: '' },
    zip: { type: String, default: '' },
    websiteUrl: { type: String, default: '' },
    faxNumber: { type: String, default: '' },
    phoneNumber: { type: String, default: '' },
    email: String,
    techConf: {
      readType: { type: String, default: 'PRELIM' },
      contrastRequirements: [
        {
          modality: { type: String },
          requirement: { type: String },
        },
      ],
    },
    dicomDetails: {
      AETitle: { type: String, default: '' },
    },
    // physicians: [{ type: ObjectId, ref: 'Physician'}],

    /*status: { type: String, default: 1, enum: [0, 1] },
      user: [{ type: ObjectId, required: true, ref: 'Users' }],
      logoUrl: { type: String, default: '' },
      note: { type: String, default: '' },
      AETitle: { type: String, default: '' },
      priorStudy: { type: Boolean, default: false },
      history: { type: Boolean, default: false },
      historyNote: { type: String, default: '' },
      retention: { type: Number, default: 90 },
      laterality: { type: String, default: '' },
      contrastStatus: { type: String, default: '' },
      priorStatus: { type: Number, default: 5 },
      studyType: { type: String, default: 'Final' },
      indication: { type: Boolean, default: false },*/
/*},
  { timestamps: true },
);

mongoose.model('Facility', Facility, 'facility');
*/
