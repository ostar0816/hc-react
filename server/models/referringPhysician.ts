'use strict';

import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import { GQLContext } from '../utils/getGraphqlContext';

export type MReferringPhysician = {
  _id: ObjectId;
  name: string;
  physicianUid: string;
  fax: string;
  phone: string;
  email: string;
  dicomValue: string;
  facilityId: ObjectId;
};

export function cReferringPhysician() {
  return getDb().collection<MReferringPhysician>('referringPhysician');
}

export async function loadReferringPhysicianByIds(ctx: GQLContext, referringPhysicianIds) {
  checkPerms(ctx);
  return ctx.loaders.ReferringPhysician.loadMany(referringPhysicianIds);
}

export async function addReferringPhysician(
  ctx: GQLContext,
  referringPhysician: Omit<MReferringPhysician, '_id'>,
) {
  checkPerms(ctx);

  return cReferringPhysician().insertOne(referringPhysician);
}

export async function updateReferringPhysician(
  ctx: GQLContext,
  referringPhysicianId: ObjectId,
  referringPhysician: Omit<MReferringPhysician, '_id' | 'facilityId'>,
) {
  checkPerms(ctx);

  return (await cReferringPhysician().findOneAndUpdate(
    { _id: referringPhysicianId },
    {
      $set: referringPhysician,
    },
    { returnOriginal: false },
  )).value;
}

export async function deleteReferringPhysician(ctx: GQLContext, referringPhysicianId: ObjectId) {
  checkPerms(ctx);

  return (await cReferringPhysician().findOneAndDelete({ _id: referringPhysicianId })).value;
}

export async function listReferringPhysiciansForFacility(ctx: GQLContext, facilityId: ObjectId) {
  checkPerms(ctx);

  return cReferringPhysician()
    .find({ facilityId })
    .sort([['name', 1]])
    .toArray();
}
