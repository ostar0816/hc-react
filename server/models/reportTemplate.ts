'use strict';

import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import { GQLContext } from '../utils/getGraphqlContext';

export type MReportTemplate = {
  _id: ObjectId;
  name: string;
  description: string;
  template: any;
};

export function cReportTemplate() {
  return getDb().collection<MReportTemplate>('reportTemplate');
}

export async function addReportTemplate(
  ctx: GQLContext,
  reportTemplate: Omit<MReportTemplate, '_id'>,
) {
  checkPerms(ctx);

  return cReportTemplate().insertOne(reportTemplate);
}

export async function updateReportTemplate(
  ctx: GQLContext,
  reportTemplateId,
  reportTemplate: Omit<MReportTemplate, '_id'>,
) {
  checkPerms(ctx);

  return (await cReportTemplate().findOneAndUpdate(
    { _id: reportTemplateId },
    {
      $set: reportTemplate,
    },
    { returnOriginal: false },
  )).value;
}

export async function deleteReportTemplate(ctx: GQLContext, reportTemplateId: ObjectId) {
  checkPerms(ctx);

  return cReportTemplate().findOneAndDelete({ _id: reportTemplateId });
}

export async function getReportTemplateById(ctx: GQLContext, reportTemplateId: ObjectId) {
  checkPerms(ctx);

  return cReportTemplate().findOne({ _id: reportTemplateId });
}

export async function listReportTemplates(ctx: GQLContext) {
  checkPerms(ctx);

  return cReportTemplate()
    .find({})
    .sort([['name', 1]])
    .toArray();
}
