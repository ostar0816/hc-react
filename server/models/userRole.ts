'use strict';
import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import { GQLPermissionEnum } from '../graphql/types';
import { GQLContext } from '../utils/getGraphqlContext';

export type MUserRole = {
  _id: ObjectId;
  name: string;
  // some roles needs static name so can be associated with users automatically
  builtInName: string;
  description: string;
  permissions: Array<GQLPermissionEnum>;
};

export function getUserRoleDefaults() {
  return {
    builtInName: '',
    permissions: [],
  };
}

export function cUserRole() {
  return getDb().collection<MUserRole>('userRole');
}

export async function getPermissions({ user }) {
  const userRole = await cUserRole().findOne({ _id: user.userRoleId });

  return userRole.permissions;
}

export async function loadUserRoleById(ctx: GQLContext, id) {
  checkPerms(ctx);
  return ctx.loaders.UserRole.load(id);
}

export async function addUserRole(
  ctx: GQLContext,
  userRole: Pick<MUserRole, 'name' | 'description'>,
) {
  checkPerms(ctx);

  return cUserRole().insertOne({ ...getUserRoleDefaults(), ...userRole });
}

export async function updateUserRole(
  ctx: GQLContext,
  userRoleId: ObjectId,
  userRole: Pick<MUserRole, 'name' | 'description'>,
) {
  checkPerms(ctx);

  return (await cUserRole().findOneAndUpdate(
    { _id: userRoleId },
    {
      $set: userRole,
    },
    {
      returnOriginal: false,
    },
  )).value;
}

type UserRolePermission = {
  userRoleId: ObjectId;
  permissions: Array<string>;
};
export async function updateUserRolePermissions(
  ctx: GQLContext,
  userRolePermissions: Array<UserRolePermission>,
) {
  const updates = userRolePermissions.map(async ({ userRoleId, permissions }) => {
    return (await cUserRole().findOneAndUpdate(
      { _id: userRoleId },
      {
        $set: {
          permissions,
        },
      },
      { returnOriginal: false },
    )).value;
  });

  return Promise.all(updates);
}

export async function listUserRole(ctx: GQLContext) {
  checkPerms(ctx);

  return cUserRole()
    .find({})
    .sort([['name', 1]])
    .toArray();
}
