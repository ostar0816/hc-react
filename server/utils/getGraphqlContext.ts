import provideDataloaders from './provideDataloaders';
import { cUser, MUser } from '../models/user';
import { getPermissions } from '../models/userRole';
import { GQLPermissionEnum } from '../graphql/types';
import { Request } from 'express';

export type GQLContext = {
  user: MUser;
  permissions: GQLPermissionEnum[];
  loaders: ReturnType<typeof provideDataloaders>;
  req: Request;
  _loaders: any;
};

async function getGraphqlContext(userId): Promise<GQLContext> {
  const user = userId ? await cUser().findOne({ _id: userId }) : null;

  const permissions = user ? await getPermissions({ user }) : null;
  return {
    user,
    permissions,
    // req gets overiden in graphql resolver
    req: null,
    loaders: provideDataloaders(),
    _loaders: {},
  };
}

export default getGraphqlContext;
