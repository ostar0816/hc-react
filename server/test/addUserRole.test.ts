import { graphql } from 'graphql';
import { cUserRole } from '../models/userRole';

import * as DummyData from './dummyData';
import getGraphqlContext from '../utils/getGraphqlContext';
import { schema, connectMongo, clearDatabase, disconnectMongo } from './helpers';

beforeAll(connectMongo);

beforeEach(clearDatabase);

afterAll(disconnectMongo);
describe('addUserRole', () => {
  it('add', async () => {
    await cUserRole().insertOne(DummyData.userRole[0]);

    const rootValue = {};
    const variables = {
      ...DummyData.userRole[1],
    };
    const context = { ...(await getGraphqlContext()), user: {}, permissions: [] };
    const query = /* GraphQL */ `
      mutation addUserRole(
        $name: String!
        $description: String!
      ) {
        addUserRole(
          name: $name
          description: $description
        ) {
          viewer {
            userRoles {
              name
              description
            }
          }
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context, variables);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });
});
