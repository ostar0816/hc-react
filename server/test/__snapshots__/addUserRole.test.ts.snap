// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUserRole add 1`] = `
Object {
  "addUserRole": Object {
    "viewer": Object {
      "userRoles": Array [
        Object {
          "description": "admin guy",
          "name": "Admin",
        },
        Object {
          "description": "rad guy",
          "name": "Radiologist",
        },
      ],
    },
  },
}
`;
